src/librustc_mir/build/expr/into.rs

```diff
@@ -163,18 +163,18 @@ impl<'a,'tcx> Builder<'a,'tcx> {
                         body_block = loop_block;
                     }
 
-                    // execute the body, branching back to the test
-                    // We write body’s “return value” into the destination of loop. This is fine,
-                    // because:
-                    //
-                    // * In Rust both loop expression and its body are required to have `()`
-                    //   as the “return value”;
-                    // * The destination will be considered uninitialised (given it was
-                    //   uninitialised before the loop) during the first iteration, thus
-                    //   disallowing its use inside the body. Alternatively, if it was already
-                    //   initialised, the `destination` can only possibly have a value of `()`,
-                    //   therefore, “mutating” the destination during iteration is fine.
-                    let body_block_end = unpack!(this.into(destination, body_block, body));
+                    // The “return” value of the loop body must always be an unit, but we cannot
+                    // reuse that as a “return” value of the whole loop expressions, because some
+                    // loops are diverging (e.g. `loop {}`). Thus, we introduce a unit temporary as
+                    // the destination for the loop body and assign the loop’s own “return” value
+                    // immediately after the iteration is finished.
+                    let tmp = this.get_unit_temp();
+                    // Execute the body, branching back to the test.
+                    let body_block_end = unpack!(this.into(&tmp, body_block, body));
+                    // Even if we assign unconditionally, if the loop turns out to be diverging,
+                    // the loop’s “return” value will not appear in the MIR (removed along with the
+                    // exit_block by cfg simplification passes later).
+                    this.cfg.push_assign_unit(exit_block, expr_span, destination);
```

Comment:
> I’m not sure if having that ill-typedness for a little while in an unreachable block is fine. 

I never answered this directly -- I'd prefer to avoid it, since I plan to be running various type-checking steps on the MIR before we do optimizations and cleanup.